libraries/standard/mqtt/src/iot_mqtt_operation.c

```diff
@@ -301,6 +332,158 @@ static bool _scheduleNextRetry( _mqttOperation_t * pOperation )
 
 /*-----------------------------------------------------------*/
 
+static bool _sendKeepAlivePingReq( _mqttConnection_t * pMqttConnection )
+{
+    bool status = true;
+    uint32_t swapStatus = 0;
+    size_t bytesSent = 0;
+    _mqttOperation_t * pPingreqOperation = &( pMqttConnection->pingreq );
+
+    /* Swap status is not checked when asserts are disabled. */
+    ( void ) swapStatus;
+
+    /* Because PINGREQ may be used to keep the MQTT connection alive, it is
+     * more important than other operations. Bypass the queue of jobs for
+     * operations by directly sending the PINGREQ in this job. */
+    bytesSent = pMqttConnection->pNetworkInterface->send( pMqttConnection->pNetworkConnection,
+                                                          pPingreqOperation->u.operation.pMqttPacket,
+                                                          pPingreqOperation->u.operation.packetSize );
+
+    if( bytesSent != pPingreqOperation->u.operation.packetSize )
+    {
+        IotLogError( "(MQTT connection %p) Failed to send PINGREQ.", pMqttConnection );
+        status = false;
+    }
+    else
+    {
+        /* Assume the keep-alive will fail. The network receive callback will
+         * clear the failure flag upon receiving a PINGRESP. */
+        swapStatus = Atomic_CompareAndSwap_u32( &( pPingreqOperation->u.operation.periodic.ping.failure ),
+                                                1,
+                                                0 );
+        IotMqtt_Assert( swapStatus == 1 );
+
+        /* Schedule a check for PINGRESP. */
+        pPingreqOperation->u.operation.periodic.ping.nextPeriodMs = IOT_MQTT_RESPONSE_WAIT_MS;
+
+        IotLogDebug( "(MQTT connection %p) PINGREQ sent. Scheduling check for PINGRESP in %d ms.",
```

Comment:
Where is the call to task pool Schedule?  This log message seems misplaced.

Reply:
The scheduling happens at line 820 in `_IotMqtt_ProcessKeepAlive`, from which the code for this function was taken. Since that also logs `"(MQTT connection %p) Next keep-alive job in %lu ms."`, I think the reference to scheduling in this statement can just be removed to make more sense.